### ISSUE TITLE
Sample different weights for examples within a  batch

### ISSUE THREAD

**Issue by vlasenkov** [tensorflow/probability#153]
If I'm not mistking currently bayesian Conv2d samples single kernel for all examples within a batch. In principle sampling different kernels for each of the examples could make MC estimates better. This requires bactched convolution where the kernel has batch dimention too. Possible approach is here:
https://stackoverflow.com/questions/42068999/tensorflow-convolutions-with-different-filter-for-each-sample-in-the-mini-batch
Does it worth to implement such feature?

**Comment by dustinvtran** [comment]
That's a good idea. Currently, we implement different Monte Carlo estimators as different `tfp.layers`. I could see such an implementation as either a new layer or as an arg to `tfp.layers.*Reparameterization`.

In practice, I don't know if this warrants much usage—[Flipout](https://openreview.net/forum?id=rJNpifWAb) was specifically designed to efficiently handle on per-example weight samples. But maybe it's useful to have anyways.

**Comment by xht033** [comment]
> That's a good idea. Currently, we implement different Monte Carlo estimators as different `tfp.layers`. I could see such an implementation as either a new layer or as an arg to `tfp.layers.*Reparameterization`.
> 
> In practice, I don't know if this warrants much usage—[Flipout](https://openreview.net/forum?id=rJNpifWAb) was specifically designed for this purpose. But maybe it's useful to have the naive version anyways. It's more general than Flipout in any case.

Does this mean that it's alright to sample the posterior distribution only once in both training and test step if we choose Flipout, however, if we choose other Bayesian layers like LocalReprameterization layer we still need to sample different posterior distribution many times?

**Comment by brianwa84** [comment]
@jvdillon the stack overflow link for conv2d is interesting here, suggests a way of having per batch member conv kernels. (I haven't verified equality)

**Comment by DimitrisCC** [comment]
I have that issue as well.
For Dense layers, I just augment the input like this: [num_particles] + input.shape,
and for the weights, using the reparam trick: W = w_mu + w_std*tf.random_normal([num_particles] + [dim_in, dim_out]),
where w_mu and w_std are variables of shape [dim_in, dim_out].
Then a tf.matmul(input_aug, W) does the trick.

But I want the same effect on 2d convolutions. I just need that: a kernel variable of [num_particles] + kernel.shape size, not everything else around it as in Conv2DReparameterization.
Seems that "faking" the extra dim with conv3d isn't working. That stackoverflow link is interesting.

Anyone got any tested ideas?

**Comment by DimitrisCC** [comment]
I have a similar issue. 
For Dense layers, I just augment the input like this: [num_particles] + input.shape,
and for the weights, using the reparam trick: W = w_mu + w_std*tf.random_normal([num_particles] + [dim_in, dim_out]),
where w_mu and w_std are trainable variables of shape [dim_in, dim_out].
Then a tf.matmul(input_aug, W) does the trick (the output maintains the num_particles dim).
So basically, it's the same weights distribution for the minibatch, but multiple different samples of it.

I want the same thing, but on 2d convolutions. 
Seems that "faking" the extra dim with tf.nn.conv3d isn't working as intended. That stackoverflow link is interesting.

Anyone got any tested ideas?

**Comment by brianwa84** [comment]
https://www.tensorflow.org/probability/api_docs/python/tfp/experimental/nn/Convolution
I
think can help here.
Internally (AIUI) it uses im2row to convert the convolution into a batch
matmul.
It's memory inefficient, and we'd be happy to have a better performance
story if someone can propose a faster approach.

Note that it takes a batch_shape argument which can serve your
[num_particles] purpose.

**Comment by DimitrisCC** [comment]
Thanks for suggestion @brianwa84! Didn't know about that function. Interesting.

Now I was trying using tf.map_fn to apply the convolution to each particle separately (does this run in parallel in GPU or something?).

So, you mean that batch_shape can be the [num_particles] dimension, filter_shape can be the kernel size (eg 3x3), input_size the in_channels and output_size the out_channels?

**Comment by DimitrisCC** [comment]
I don't want though the "augmented" kernel to be itself learnable, but only the mean and standard deviation of it which are not "augmented" with particles (so standard 4D filters). Then the "augmented" kernel is created by the param trick using this mean and std and random normal samples from the "augmented" space (~Normal([num_particles]+kernel_size)). So it's like the [num_particles] dim is not learnable (?).

Maybe I can get ideas from this implementation though, like re-arranging the dims to make it a simple tf.matmul.

**Comment by brianwa84** [comment]
Sounds like you want the num_particles axis of the kernel to be broadcast, and instead to draw multiple samples. @jvdillon is it possible?

**Comment by DimitrisCC** [comment]
Yes, exactly. Conceptually, I want multiple kernel samples from a Normal(k_mu, k_std), where k_mu and k_std are standard 4D learnable variables/filter, and do this with reparameterization trick. Then apply convolution to my input and end up with num_particles convolved samples.

In the Dense case, by simply augmenting the input with another dim and tf.matmul-ing it works.
In the Conv2D case, after the reparam trick, I don't know how to apply the convolution directly.

It's like I want a bare-bones Conv2DReparameterization, but with multiple samples as the output.
